src/gateway.rs

```diff
@@ -100,43 +108,55 @@ impl Gateway {
             self.udp_runtime
                 .prepare_empty_downlink(downlink.gateway_mac),
         );
-        let pull_resp = downlink.to_pull_resp(false)?;
-        if pull_resp.is_none() {
-            return Ok(());
-        }
-        info!(logger, "sending rx1 downlink {:?}", pull_resp);
-        downlink_rx1.set_packet(pull_resp.unwrap());
-        match downlink_rx1
-            .dispatch(Some(Duration::from_secs(DOWNLINK_TIMEOUT_SECS)))
-            .await
-        {
-            // On a too early or too late error retry on the rx2 slot if available.
-            Err(SemtechError::Ack(tx_ack::Error::TOO_EARLY))
-            | Err(SemtechError::Ack(tx_ack::Error::TOO_LATE)) => {
-                if let Some(pull_resp) = downlink.to_pull_resp(true)? {
-                    info!(logger, "sending rx2 downlink {:?}", pull_resp);
-                    downlink_rx2.set_packet(pull_resp);
-                    match downlink_rx2
-                        .dispatch(Some(Duration::from_secs(DOWNLINK_TIMEOUT_SECS)))
-                        .await
-                    {
-                        Err(SemtechError::Ack(tx_ack::Error::NONE)) => Ok(()),
-                        Err(err) => {
-                            warn!(logger, "ignoring rx2 downlink error: {:?}", err);
+
+        match downlink.to_pull_resp(false)? {
+            None => return Ok(()),
+            Some(txpk) => {
+                info!(
+                    logger,
+                    "rx1 downlink {} via {}",
+                    txpk,
+                    downlink_rx1.get_destination_mac()
+                );
+                downlink_rx1.set_packet(txpk);
+                match downlink_rx1
+                    .dispatch(Some(Duration::from_secs(DOWNLINK_TIMEOUT_SECS)))
+                    .await
+                {
+                    // On a too early or too late error retry on the rx2 slot if available.
+                    Err(SemtechError::Ack(tx_ack::Error::TOO_EARLY))
+                    | Err(SemtechError::Ack(tx_ack::Error::TOO_LATE)) => {
+                        if let Some(txpk) = downlink.to_pull_resp(true)? {
+                            info!(
+                                logger,
+                                "rx2 downlink {} via {}",
+                                txpk,
+                                downlink_rx2.get_destination_mac()
+                            );
+                            downlink_rx2.set_packet(txpk);
+                            match downlink_rx2
+                                .dispatch(Some(Duration::from_secs(DOWNLINK_TIMEOUT_SECS)))
+                                .await
+                            {
+                                Err(SemtechError::Ack(tx_ack::Error::NONE)) => Ok(()),
```

Comment:
wut.. what is an Error::NONE but an Ok in disguise?

Reply:
This was your code! :P

but yes, I can try to fix this upstream but let's do in another PR? I think this is an artifact of lazy GWMP deserializing

Reply:
It was indeed, but since I have you :-)

Reply:
oh hey i think its fixed already kind of

Reply:
uhh how? So many Error::NONE to remove?

Reply:
Dispatch returns error by asking `ack.get_error()`

https://github.com/helium/semtech-udp/blob/master/src/server_runtime/mod.rs#L107

which only returns an error if there it's not `NO_ERROR`

https://github.com/helium/semtech-udp/blob/master/src/packet/tx_ack.rs#L30

Reply:
nice! Want to adjust this PR?

Reply:
I'm looking at it... the NO_ERROR still gets kicked back and would require handling at this level.

This is the best I can do without improving the deserializer:
```
Ok(()) | Err(SemtechError::Ack(tx_ack::Error::NONE)) => Ok(()),
```

Reply:
I'm looking at deserializer improvements to see if i can swing it

Reply:
[Changes made in `semtech-udp`](https://github.com/helium/semtech-udp/pull/33) to improve the `tx_ack::Error` type. Release `0.6.0` includes it